README.md
# LogPage [![NPM version](https://img.shields.io/npm/v/logpage.svg?style=flat)](https://npmjs.com/package/logpage)

Helps you quickly deploy a live updated Changelog page based on a `CHANGELOG.md` file, inspired by [headwayapp.co](https://headwayapp.co/), [LogLive](https://github.com/egoist/loglive).

<img src="https://i.loli.net/2017/07/27/597995184bb02.png" width="800" alt="preview">

Live Demo: https://logp.netlify.app

## Usage

You can simply create an `index.html` and include `logpage` there:

```html
<html>

<head>
    <title>LogPage Example</title>
    <link rel="stylesheet" href="https://unpkg.com/logpage/dist/style.css" />
</head>

<body>
    <div id="app"></div>
    <script type="module">
        import LogPage from 'https://unpkg.com/logpage';
        new LogPage({
            target: document.getElementById('app')
        });
    </script>
</body>

</html>
```

Then you can start to write a `CHANGELOG.md`, by default LogPage will read from `http://your-website/CHANGELOG.md`, however you can use a custom path:

```
new LogPage({
  target: document.getElementById('app'),
  props: {}
});
```

## Config

- `showHeader` (boolean): Show the header or not

## Changelog Format

### Example

```md
# My Website

## Add TypeScript support (2016-07-23)

### New

- Now added Typescript support.
- Some other new stuffs.

### Fix

Fixed some babel problems.
```

### Site name

The value of `h1` heading will be used as website name:

```md
# My Website
```

### Changelog Title

Changelog title is an `h2` heading which consists of two parts, `title` and `date` and `date` is totally optional.

```md
## changelog title (date)
```

Sometime you just want to use semantic version as the changelog title:

```md
## 1.0.0 (2023-06-12)
```

### Change Type

Change type is an `h3` heading which describes the type of the change, eg: `Fix`. And it shows up like:

<img src="https://i.loli.net/2017/07/27/59797da5a89df.png" alt="change type" width="60">

By default we have pre-defined colors for specific types which include: `Fix` `Breaking` `New`, other types will get a random color.

You can also set color for custom change type:

```js
new LogPage({
    target: document.getElementById('app'),
    props: {
        colors: {
            'Bug fixes': '#342343'
        }
    }
});
```

## Front Matters

### website

You changelog is born for your product, so adding a link to your actual product is necessary:

```md
---
website: http://my-fantastic-app.com
---
```


CHANGELOG.md
---
website: https://github.com/isaced/logpage
---

# LogPage

## 0.0.2 (2023-06-12T14:09:50.125Z)

### New

- Add support dark mode
- Add support for custom hidden header

```javascript
new LogPage({
    target: document.getElementById('app'),
    props: {
        showHeader: false, // controls whether to show the header
    }
});
```

## 0.0.1 (2023-06-12T05:52:25.142Z)

### Initial

I'm just publising it to see if it works as expected, it still has a lot of spaces to improve but I feel this version would work just fine.


src/utils/renderer.js
import marked from 'marked3'
import { formatDistanceToNow, parseISO } from 'date-fns'

const COLORS = [
  '#9e0734',
  '#016b18',
  '#4e6d00',
  '#42890c',
  '#457c0b',
  '#d1b908',
  '#a000d1',
  '#458704',
  '#0a1172',
  '#4c0172',
  '#377705',
  '#10a870',
  '#047531',
  '#870004',
  '#149903',
  '#01722c',
  '#073870',
  '#0f0d87',
  '#bf07bf',
  '#af0002'
]

const randomColor = () => {
  return COLORS[Math.floor(Math.random() * 20)]
}

const TITLE_REGEX = /^([^(]+)(?:\(([^)]+)\))?$/

const matchChangelogTitle = title => {
  return TITLE_REGEX.exec(title)
}

export default ({ site, colors }) => {
  const renderer = new marked.Renderer()

  const originalHeading = renderer.heading.bind(renderer)

  renderer.heading = (text, level, raw) => {
    if (level === 1) {
      site.title = text
      return ''
    }

    if (level === 2) {
      const matched = matchChangelogTitle(raw)
      if (matched) {
        let [, title, date = ''] = matched
        const dateParsed = date && parseISO(date)
        title = title.trim()
        return `<div class="logpage-logtitle">${originalHeading(
          title,
          level,
          title
        )}<time class="logpage-date" datetime="${dateParsed}">${date && formatDistanceToNow(dateParsed, { addSuffix: true })}</time></div>`
      }
    }

    if (level === 3) {
      let res = '<div class="logpage-sep"></div>'
      const color = colors[raw.trim()] || randomColor()
      res +=
        originalHeading(text, level, raw).replace(
          /^<h3/,
          `<h3 style="background-color: ${color}" class="logpage-h3"`
        ) + '<div class="clearfix"></div>'
      return res
    }

    return originalHeading(text, level, raw)
  }

  return renderer
}


src/utils/renderer.ts
import { Renderer } from "marked";
import { formatDistanceToNow, parseISO } from "date-fns";

const COLORS = [
  "#9e0734",
  "#016b18",
  "#4e6d00",
  "#42890c",
  "#457c0b",
  "#d1b908",
  "#a000d1",
  "#458704",
  "#0a1172",
  "#4c0172",
  "#377705",
  "#10a870",
  "#047531",
  "#870004",
  "#149903",
  "#01722c",
  "#073870",
  "#0f0d87",
  "#bf07bf",
  "#af0002",
];

const randomColor = () => {
  return COLORS[Math.floor(Math.random() * 20)];
};

const TITLE_REGEX = /^([^(]+)(?:\(([^)]+)\))?$/;

const matchChangelogTitle = (title) => {
  return TITLE_REGEX.exec(title);
};

export default ({ site, colors }) => {
  const renderer = new Renderer();

  const originalHeading = renderer.heading.bind(renderer);

  renderer.heading = (text, level, raw, slugger) => {
    if (level === 1) {
      site.title = text;
      return "";
    }

    if (level === 2) {
      const matched = matchChangelogTitle(raw);
      if (matched) {
        let [, title, date = ""] = matched;
        const dateParsed = date && parseISO(date);
        title = title.trim();
        return `<div class="logpage-logtitle">${originalHeading(
          title,
          level,
          title,
          slugger
        )}<time class="logpage-date" datetime="${dateParsed}">${
          date && formatDistanceToNow(dateParsed, { addSuffix: true })
        }</time></div>`;
      }
    }

    if (level === 3) {
      let res = '<div class="logpage-sep"></div>';
      const color = colors[raw.trim()] || randomColor();
      res +=
        originalHeading(text, level, raw, slugger).replace(
          /^<h3/,
          `<h3 style="background-color: ${color}" class="logpage-h3"`
        ) + '<div class="clearfix"></div>';
      return res;
    }

    return originalHeading(text, level, raw, slugger);
  };

  return renderer;
};
